### PR TITLE
executor, infoschema: fix display of on update CURRENT_TIMESTAMP with decimal

### LINKED ISSUE #11349: dump of `on update` missing fraction
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
create table xyz using:
```
CREATE TABLE `xyz` (
  `id` varchar(36) NOT NULL,
  `create_at` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
  `update_at` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3),
  PRIMARY KEY (`id`)
);
```

then using mysql dump to extract structure.

2. What did you expect to see（result from `circleci/mysql:5.7`）?

```
SET NAMES utf8mb4;
SET FOREIGN_KEY_CHECKS = 0;

-- ----------------------------
-- Table structure for xyz
-- ----------------------------
DROP TABLE IF EXISTS `xyz`;
CREATE TABLE `xyz` (
  `id` varchar(36) COLLATE utf8_unicode_ci NOT NULL,
  `create_at` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
  `update_at` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3),
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;

SET FOREIGN_KEY_CHECKS = 1;
```

3. What did you see instead(result from TiDB)?

```
SET NAMES utf8mb4;
SET FOREIGN_KEY_CHECKS = 0;

-- ----------------------------
-- Table structure for xyz
-- ----------------------------
DROP TABLE IF EXISTS `xyz`;
CREATE TABLE `xyz` (
  `id` varchar(36) NOT NULL,
  `create_at` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
  `update_at` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

SET FOREIGN_KEY_CHECKS = 1;
```

The `ON UPDATE CURRENT_TIMESTAMP` missing fraction

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
3.0.1

5. Addon

`explicit-defaults-for-timestamp` has set ON

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve? 
Fix #11349.

### What is changed and how it works?
1. refactor `IsCurrentTimestampExpr` into `IsValidCurrentTimestampExpr`. Here "valid" means the fractional seconds precision in column type equals to the one in `on update` clause.
2. print fractional seconds precision in `on update` clause when `show create table` and `describe` if necessary. For example, here is a table with the following schema:
```sql
create table t (
  a datetime not null default current_timestamp on update current_timestamp,
  b datetime(3) not null default current_timestamp(3) on update current_timestamp(3)
);
```

show create table t:
```sql
+-------+----------------------------------------------------------------------------------------+
| Table | Create Table                                                                           |
+-------+----------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                                                     |
|       |   `a` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,         |
|       |   `b` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3) |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin                            |
+-------+----------------------------------------------------------------------------------------+
```

desc t:
```sql
+-------+-------------+------+-----+----------------------+--------------------------------------------------+
| Field | Type        | Null | Key | Default              | Extra                                            |
+-------+-------------+------+-----+----------------------+--------------------------------------------------+
| a     | datetime    | NO   |     | CURRENT_TIMESTAMP    | DEFAULT_GENERATED on update CURRENT_TIMESTAMP    |
| b     | datetime(3) | NO   |     | CURRENT_TIMESTAMP(3) | DEFAULT_GENERATED on update CURRENT_TIMESTAMP(3) |
+-------+-------------+------+-----+----------------------+--------------------------------------------------+
```

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: expression/helper.go [hunk 1]
```diff
@@ -34,12 +34,23 @@ func boolToInt64(v bool) int64 {
...
+// Here `valid` means it is consistent with the given fieldType's Decimal.
+func IsValidCurrentTimestampExpr(exprNode ast.ExprNode, fieldType *types.FieldType) bool {
+	fn, isFuncCall := exprNode.(*ast.FuncCallExpr)
+	if !isFuncCall || fn.FnName.L != ast.CurrentTimestamp {
+		return false
 	}
-	return false
+
+	containsArg := len(fn.Args) > 0
+	containsFsp := fieldType != nil && fieldType.Decimal > 0 // Fsp represents fractional seconds precision.
```
Comment (AilinKid) [inline expression/helper.go]: Comment is better in a new line

File: expression/helper.go [hunk 2]
```diff
@@ -34,12 +34,23 @@ func boolToInt64(v bool) int64 {
...
+	var isConsistent bool
+	if containsArg {
+		v, ok := fn.Args[0].(*driver.ValueExpr)
+		isConsistent = ok && fieldType != nil && v.Datum.GetInt64() == int64(fieldType.Decimal)
+	}
+
+	return (containsArg && isConsistent) || (!containsArg && !containsFsp)
```
Comment (AilinKid) [inline expression/helper.go]: Could we test when it is not valid?

File: expression/helper.go (hunk 1 above)
Comment (tangenta) [reply expression/helper.go]: @AilinKid Updated. PTAL

File: expression/helper.go (hunk 2 above)
Comment (tangenta) [reply expression/helper.go]: @AilinKid Added integration tests for invalid cases.

File: expression/helper.go [hunk 3]
```diff
@@ -34,12 +34,24 @@ func boolToInt64(v bool) int64 {
 	return 0
 }
 
-// IsCurrentTimestampExpr returns whether e is CurrentTimestamp expression.
-func IsCurrentTimestampExpr(e ast.ExprNode) bool {
-	if fn, ok := e.(*ast.FuncCallExpr); ok && fn.FnName.L == ast.CurrentTimestamp {
-		return true
+// IsValidCurrentTimestampExpr returns whether e is a valid CurrentTimestamp expression.
```
Comment (winkyao) [inline expression/helper.go]: whether `exprNode`

Comment (tangenta) [reply expression/helper.go]: Updated. PTAL

File: ddl/db_integration_test.go [hunk 4]
```diff
@@ -416,6 +416,14 @@ func (s *testIntegrationSuite5) TestMySQLErrorCode(c *C) {
 	assertErrorCode(c, tk, sql, tmysql.ErrPrimaryCantHaveNull)
 	sql = "create table t2 (id int auto_increment);"
 	assertErrorCode(c, tk, sql, tmysql.ErrWrongAutoKey)
+	sql = "create table t2 (a datetime(2) default current_timestamp(3))"
```
Comment (bb7133) [inline ddl/db_integration_test.go]: Please check the following statements and add related test cases:

```
mysql> show columns from t;
+-------+-------------+------+-----+----------------------+--------------------------------+
| Field | Type        | Null | Key | Default              | Extra                          |
+-------+-------------+------+-----+----------------------+--------------------------------+
| a     | datetime    | NO   |     | CURRENT_TIMESTAMP    | on update CURRENT_TIMESTAMP    |
| b     | datetime(3) | NO   |     | CURRENT_TIMESTAMP(3) | on update CURRENT_TIMESTAMP(3) |
+-------+-------------+------+-----+----------------------+--------------------------------+
2 rows in set (0.01 sec)

mysql> select * from information_schema.columns where table_name = 't' and column_name='b'\G
*************************** 1. row ***************************
           TABLE_CATALOG: def
            TABLE_SCHEMA: test
              TABLE_NAME: t
             COLUMN_NAME: b
        ORDINAL_POSITION: 2
          COLUMN_DEFAULT: CURRENT_TIMESTAMP(3)
             IS_NULLABLE: NO
               DATA_TYPE: datetime
CHARACTER_MAXIMUM_LENGTH: NULL
  CHARACTER_OCTET_LENGTH: NULL
       NUMERIC_PRECISION: NULL
           NUMERIC_SCALE: NULL
      DATETIME_PRECISION: 3
      CHARACTER_SET_NAME: NULL
          COLLATION_NAME: NULL
             COLUMN_TYPE: datetime(3)
              COLUMN_KEY: 
                   EXTRA: on update CURRENT_TIMESTAMP(3)
              PRIVILEGES: select,insert,update,references
          COLUMN_COMMENT: 
   GENERATION_EXPRESSION: 
1 row in set (0.01 sec)

```

Comment (tangenta) [reply ddl/db_integration_test.go]: Updated. PTAL @bb7133

File: ddl/db_integration_test.go [hunk 5]
```diff
@@ -1907,3 +1907,21 @@ func (s *testIntegrationSuite4) TestDropAutoIncrementIndex(c *C) {
...
+	tk.MustExec("use test_on_update_timestamp")
+	tk.MustExec("create table timestamp_table (a datetime not null default current_timestamp on update current_timestamp, " +
+		"b datetime(3) not null default current_timestamp(3) on update current_timestamp(3))")
+	tk.MustQuery("show full columns from timestamp_table").Check(testkit.Rows(
+		"a datetime <nil> NO  CURRENT_TIMESTAMP DEFAULT_GENERATED on update CURRENT_TIMESTAMP select,insert,update,references ",
+		"b datetime(3) <nil> NO  CURRENT_TIMESTAMP(3) DEFAULT_GENERATED on update CURRENT_TIMESTAMP(3) select,insert,update,references "))
+	tk.MustQuery("select * from information_schema.columns where table_name = 'timestamp_table' and column_name='a'").Check(testkit.Rows(
+		"def test_on_update_timestamp timestamp_table a 1 CURRENT_TIMESTAMP NO datetime <nil> <nil> <nil> <nil> 0 " +
+			"<nil> <nil> datetime  DEFAULT_GENERATED on update CURRENT_TIMESTAMP select,insert,update,references  "))
+	tk.MustQuery("select * from information_schema.columns where table_name = 'timestamp_table' and column_name='b'").Check(testkit.Rows(
```
Comment (bb7133) [inline ddl/db_integration_test.go]: Please move the related tests here:
https://github.com/pingcap/tidb/blob/d2b5387183a591847f952aa4e3a3c4597c34dcba/infoschema/tables_test.go#L262

File: ddl/db_integration_test.go [hunk 6]
```diff
@@ -1907,3 +1907,21 @@ func (s *testIntegrationSuite4) TestDropAutoIncrementIndex(c *C) {
...
 }
+
+func (s *testIntegrationSuite4) TestOnUpdateCurrentTimestamp(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("drop database if exists test_on_update_timestamp")
+	tk.MustExec("create database test_on_update_timestamp")
+	tk.MustExec("use test_on_update_timestamp")
+	tk.MustExec("create table timestamp_table (a datetime not null default current_timestamp on update current_timestamp, " +
+		"b datetime(3) not null default current_timestamp(3) on update current_timestamp(3))")
+	tk.MustQuery("show full columns from timestamp_table").Check(testkit.Rows(
```
Comment (bb7133) [inline ddl/db_integration_test.go]: Please move the "show columns" tests to here:
https://github.com/pingcap/tidb/blob/d2b5387183a591847f952aa4e3a3c4597c34dcba/executor/show_test.go#L165

**Pull request merged**